src/Val.h

```diff
@@ -506,38 +343,117 @@ class ValManager {
 
 extern ValManager* val_mgr;
 
+
+// These are *internal* classes used to allow different publicly visible
+// classes to share the same low-level value (per Type::InternalType).
+// They may change or go away in the future.
+
+class IntValImplementation : public Val {
+public:
+	IntValImplementation(TypePtr t, bro_int_t v)
+		: Val(std::move(t)), int_val(v)
+		{}
+
+	bro_int_t Get() const	{ return int_val; }
+
+protected:
+	bro_int_t int_val;
+};
+
+class UnsignedValImplementation : public Val {
+public:
+	UnsignedValImplementation(TypePtr t, bro_uint_t v)
+		: Val(std::move(t)), uint_val(v)
+		{}
+
+	bro_uint_t Get() const	{ return uint_val; }
+
+protected:
+	bro_uint_t uint_val;
+};
+
+class DoubleValImplementation : public Val {
+public:
+	DoubleValImplementation(TypePtr t, double v)
+		: Val(std::move(t)), double_val(v)
+		{}
+
+	double Get() const	{ return double_val; }
+
+protected:
+	double double_val;
+};
+
+
+class IntVal final : public IntValImplementation {
```

Comment:
Actually I had thought we could use double inheritance, e.g.:

```
class IntVal final : public Val, public IntValImplementation
```

(where `DoubleValImplementation` would not inherit from anything). Not a huge difference, but would make it a bit clearer what's going on (unless I'm missing a reason why that wouldn't work).